### PR TITLE
Update the state machine to wait for machineType before kicking off the init process

### PR DESCRIPTION
**Pull request by clholgat**
Ticket: https://github.com/AllYarnsAreBeautiful/ayab-firmware/issues/52
Desktop change: https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/447

This is one of those, how did that ever work, kind of things. Looks like this was something that got lost in the move to monolithic firmware.

Adds a new starting state to the FSM that waits for the desktop software to notify if of the selected machine type before proceeding.

This touched a lot of tests.

### CODE REVIEW

**Comment by t0mpr1c3** [comment]
I guess this relates to #52? Do we need corresponding changes in the desktop?

**Comment by clholgat** [comment]
Yes, this is the firmware side of that fix. I uploaded a draft so I could see the diff. There will be a desktop change and tests coming. I'll let y'all know when it's ready for review.

File: src/ayab/encoders.h
Comment (dl1com) [inline src/ayab/encoders.h:1]: These changes seem not to be related to the update to the state machine structure

File: src/ayab/encoders.cpp
Comment (dl1com) [inline src/ayab/encoders.cpp:1]: These changes seem not to be related to the update to the state machine structure

Comment (clholgat) [reply src/ayab/encoders.cpp:1]: These are garter carriage fixes, I can put them in a separate PR.

File: src/ayab/com.cpp [hunk 1]
```diff
@@ -101,6 +101,10 @@ void Com::send_indState(Carriage_t carriage, uint8_t position,
  */
 void Com::onPacketReceived(const uint8_t *buffer, size_t size) {
   switch (buffer[0]) {
+  case reqInit_msgid:
```
Comment (jpcornil-git) [inline src/ayab/com.cpp:104]: Why not add this parameter to reqStart ? What about adding a Shield HW version as well, e.g. ignore right Hall sensor with current HW/handle with next one ?

File: src/ayab/encoders.h [hunk 2]
```diff
@@ -63,14 +63,14 @@ constexpr uint8_t END_LEFT[NUM_MACHINES] = {0U, 0U, 0U};
...
-constexpr uint8_t GARTER_SLOP = 14;
+constexpr uint8_t GARTER_SLOP = 2;
 
 constexpr uint8_t START_OFFSET[NUM_MACHINES][NUM_DIRECTIONS][NUM_CARRIAGES] = {
     // KH910
     {
         //   K,  L,  G
-        {40, 40, 8}, // Left
-        {16, 16, 32} // Right
+        {40, 40, 31}, // Left
```
Comment (jpcornil-git) [inline src/ayab/encoders.h:72]: Where are these new offsets for G carriage coming from ?
+7 needles from what I see below
31 % 16 = 15 vs 8 % 16 = 8
55 % 16 = 7 vs 32 % 16 = 0

File: src/ayab/knitter.cpp [hunk 3]
```diff
@@ -118,34 +116,49 @@ void Knitter::isr() {
...
+    return NO_MACHINE_TYPE;
+  }
+  if (machineType >= NUM_MACHINES) {
+    return MACHINE_TYPE_INVALID;
+  }
+
+  m_machineType = machineType;
+
+  // Now that we have enough start state, we can set up interrupts
+  setUpInterrupt();
```
Comment (jpcornil-git) [inline src/ayab/knitter.cpp:133]: Should this (enable interrupts) not move after encoder/FSM init L135/136 ?

**Review by jpcornil-git — COMMENTED** [review]
Just a few comments while looking at this PR

File: src/ayab/encoders.h (hunk 2 above)
Comment (clholgat) [reply src/ayab/encoders.h:72]: They came from manual testing.  I'm going to pull them into a separate PR.

File: src/ayab/com.cpp (hunk 1 above)
Comment (jpcornil-git) [reply src/ayab/com.cpp:104]: Indeed, you need machineType if the capability to run a machine-dependent carriage detection is required ... which is the case for the right hall sensor already/OK.
Wrt cnfinfo, I see FW version but no HW version (yet ?)

File: src/ayab/encoders.h (hunk 2 above)
Comment (jpcornil-git) [reply src/ayab/encoders.h:72]: I guess that 55 would give the same behavior as 7 as everything is modulo 16 for solenoids ? (meaning "G shifted by 9 vs K/L",, i.e. 40-9 and 16-9)
As this is related to the physical position of the G magnet, central for K/L and shifted for G, I would have said 12 iso 9 for G a priori, i.e 28 and 4, as this is more or less the distance in needles between the magnet position and the center of the carriage.

**Comment by dl1com** [comment]
Please let me know when the discussions are resolved, so I can merge this together with https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/447

**Comment by jpcornil-git** [comment]
> Please let me know when the discussions are resolved, so I can merge this together with [AllYarnsAreBeautiful/ayab-desktop#447](https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/447)

Except for the remaining https://github.com/AllYarnsAreBeautiful/ayab-firmware/pull/60#discussion_r1164631206, everything else is related to separate PRs/discussions

File: src/ayab/com.cpp (hunk 1 above)
Comment (clholgat) [reply src/ayab/com.cpp:104]: I don't think we have a way of detecting HW version from the firmware yet.

File: src/ayab/encoders.h (hunk 2 above)
Comment (clholgat) [reply src/ayab/encoders.h]: I will keep that in mind.

File: src/ayab/knitter.cpp (hunk 3 above)
Comment (clholgat) [reply src/ayab/knitter.cpp]: The interrupt used to be setup long before the encoder, but yea, for consistency sake, it probably should be.

**Review by X-sam — APPROVED** [review]
looks good from here...

**Pull request merged**